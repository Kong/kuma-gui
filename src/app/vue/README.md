---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

