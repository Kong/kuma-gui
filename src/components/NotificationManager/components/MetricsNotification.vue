<template>
  <div class="py-4">
    <p class="mb-4">
      A traffic metrics policy lets you collect key data for observability of your service mesh.
    </p>

    <ul class="list-disc pl-4">
      <li>
        <a
          :href="`https://kuma.io/docs/${version}/policies/traffic-metrics/`"
          target="_blank"
        >
          Traffic Metrics policy documentation
        </a>
      </li>
    </ul>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
export default {
  name: 'MetricsNotification',
  computed: {
    ...mapGetters({
      version: 'config/getVersion',
    }),
  },
}
</script>
