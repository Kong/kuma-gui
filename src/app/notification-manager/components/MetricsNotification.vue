<template>
  <div class="py-4">
    <p class="mb-4">
      A traffic metrics policy lets you collect key data for observability of your service mesh.
    </p>

    <ul class="list-disc pl-4">
      <li>
        <a
          :href="`https://kuma.io/docs/${kumaDocsVersion}/policies/traffic-metrics/`"
          target="_blank"
        >
          Traffic Metrics policy documentation
        </a>
      </li>
    </ul>
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue'
import { useStore } from '@/store/store'

const store = useStore()

const kumaDocsVersion = computed(() => store.getters['config/getKumaDocsVersion'])
</script>
