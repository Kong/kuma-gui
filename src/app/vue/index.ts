import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
