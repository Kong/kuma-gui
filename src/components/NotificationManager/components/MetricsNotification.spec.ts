import { createStore } from 'vuex'
import { render } from '@testing-library/vue'

import MetricsNotification from './MetricsNotification.vue'

describe('MetricsNotification.vue', () => {
  it('renders snapshot', () => {
    const store = createStore({
      modules: {
        config: {
          namespaced: true,
          state: {
            kumaDocsVersion: '1.2.0',
          },
          getters: {
            getKumaDocsVersion: (state) => state.kumaDocsVersion,
          },
        },
      },
    })

    const { container } = render(MetricsNotification, {
      global: {
        plugins: [store],
      },
    })

    expect(container).toMatchSnapshot()
  })
})
