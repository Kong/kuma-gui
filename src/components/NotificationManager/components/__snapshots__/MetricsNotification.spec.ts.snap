// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsNotification.vue renders snapshot 1`] = `
<div>
  <div
    class="py-4"
  >
    <p
      class="mb-4"
    >
      
      Observability will give you a better understanding of service behavior
      to increase development teamsâ€™ reliability and efficiency.
    
    </p>
     
    <p
      class="mb-4"
    >
      
      Kuma natively integrates with 
      <a
        href="https://prometheus.io/docs/visualization/grafana/"
        target="_blank"
      >
        Prometheus
      </a>
       for auto-service discovery and traffic metrics
      collection. We also integrate with Grafana dashboards for performance monitoring. You can install your own
      Prometheus and Grafana, or you can run
      
      <div
        class="tab-container"
        data-testid="tab-container"
      >
        <!---->
         
        <div
          class="tab__content-container"
        >
          <div
            class="k-tabs"
            data-v-a38b664a=""
          >
            <ul
              aria-label="ktabs"
              data-v-a38b664a=""
              role="tablist"
            >
              <li
                aria-controls="panel-0"
                aria-selected="true"
                class="tab-item active"
                data-v-a38b664a=""
                id="universal-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Universal
                </a>
              </li>
              <li
                aria-controls="panel-1"
                aria-selected="false"
                class="tab-item"
                data-v-a38b664a=""
                id="kubernetes-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Kubernetes
                </a>
              </li>
            </ul>
            <div
              aria-labelledby="universal-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-0"
              role="tabpanel"
              tabindex="0"
            >
              <code
                class="block"
                data-v-a38b664a=""
              >
                <pre
                  data-v-a38b664a=""
                >
                  type: Mesh
name: default
metrics:
  enabledBackend: prometheus-1
  backends:
  - name: prometheus-1
    type: prometheus
    conf:
      skipMTLS: true 
                </pre>
              </code>
              
          We can apply the configuration with
          
              <code
                class="inline"
                data-v-a38b664a=""
              >
                kubectl apply -f [..]
              </code>
              .
        
            </div>
            <div
              aria-labelledby="kubernetes-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-1"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
          </div>
           
          <div>
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </p>
     
    <p
      class="mb-4"
    >
      
      If you want to use Splunk, Logstash or any other system, Kuma supports
      those as well.
    
    </p>
     
    <a
      class="k-button outline-primary"
      data-v-38fa8ecf=""
      href="https://kuma.io/docs/1.2.0/policies/traffic-metrics/"
      target="_blank"
      type="button"
    >
      
      Learn More
    
    </a>
  </div>
</div>
`;
